File: android/src/main/kotlin/com/mytiki/sdk/capacitor/TikiSdk.kt
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

package com.mytiki.sdk.capacitor

import android.content.Context
import com.getcapacitor.JSObject
import com.getcapacitor.PluginCall
import com.mytiki.tiki_sdk_android.channel.Channel
import com.mytiki.tiki_sdk_android.trail.rsp.RspInitialize
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.async

class TikiSdk {
    private val channel: Channel = Channel()
    val trail: Trail = Trail(channel)
    val idp: Idp = Idp(channel)

    private var id: String? = null
    private var address: String? = null

    fun id(call: PluginCall) {
        val ret = JSObject()
        ret.put("id", id)
        call.resolve(ret)
    }

    fun address(call: PluginCall) {
        val ret = JSObject()
        ret.put("address", address)
        call.resolve(ret)
    }

    fun isInitialized(call: PluginCall) {
        val ret = JSObject()
        ret.put("isInitialized", address != null)
        call.resolve(ret)
    }

    fun initialize(call: PluginCall, context: Context) {
        val deferred = MainScope().async {
            val rsp: RspInitialize = channel.initialize(
                call.getString("id")!!,
                call.getString("publishingId")!!,
                context
            ).await()
            id = rsp.id
            address = rsp.address
            val ret = JSObject()
            ret.put("id", id)
            ret.put("address", address)
            call.resolve(ret)
        }
    }
}

File: ios/Plugin/TikiSdk.swift
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

import Foundation
import TikiSdk
import Capacitor

@objc public class TikiSdkCore: NSObject {

    public var instance = TikiSdk.config()
    
    private var id: String?
    private var address: String?

    @objc func initialize(call: CAPPluginCall) {
            Task {
                let userId = call.getString("id")!
                let publishingId = call.getString("publishingId")!
                do{
                    try await TikiSdk.config().initialize(id: userId,  publishingId: publishingId) {
                        do{
                            self.address = try TikiSdk.address
                            self.id = userId
                            var ret = JSObject()
                            ret["id"] = self.id
                            ret["address"] = self.address
                            call.resolve(ret)
                        }catch{
                            call.reject("Tiki Sdk failed to initialize: \(error)")
                        }
                    }
                }catch{
                    call.reject("Tiki Sdk failed to initialize: \(error)")
                }
            }
    }

    @objc func getId(call: CAPPluginCall) {
        var ret = JSObject()
        ret["id"] = id
        call.resolve(ret)
    }

    @objc func getAddress(call: CAPPluginCall) {
        var ret = JSObject()
        ret["address"] = address
        call.resolve(ret)
    }

    @objc func isInitialized(call: CAPPluginCall) {
        var ret = JSObject()
        ret["isInitialized"] = address != nil
        call.resolve(ret)
    }

}
